pennylane/ops/qubit/special_unitary.py

```diff
@@ -298,32 +434,240 @@ def compute_matrix(theta, num_wires):
+
+            U(\theta) &= \exp\{A(\theta)\}\\
+            A(\theta) &= \sum_{m=1}^d i \theta_m P_m\\
+            P_m &\in \{I, X, Y, Z\}^{\otimes n} \setminus \{I^{\otimes n}\}
+
+        Then the partial derivatives of the gate can be shown to be given by
+
+        .. math::
+
+            \frac{\partial}{\partial \theta_\ell} U(\theta) &= U(\theta)
+            \frac{\mathrm{d}}{\mathrm{d}t}\exp\left(t\Omega_\ell(\theta)\right)\large|_{t=0}
+
+        where :math:`\Omega_\ell(\theta)` is the one-parameter generator belonging to the partial
+        derivative :math:`\partial_\ell U(\theta)` at the parameters :math:`\theta`.
+        It can be computed via
+
+        .. math::
+
+            \Omega_\ell(\theta) &= U(\theta)^\dagger
+            \left(\frac{\partial}{\partial \theta_\ell}\mathfrak{Re}[U(\theta)]
+            +i\frac{\partial}{\partial \theta_\ell}\mathfrak{Im}[U(\theta)]\right)
+
+        where we may compute the derivatives :math:`\frac{\partial}{\partial \theta_\ell} U(\theta)` using auto-differentiation.
 
-        .. seealso:: :meth:`~.SpecialUnitary.decomposition`.
+        .. warning::
+
+            An auto-differentiation framework is required for this function.
+            The matrix exponential is not differentiable in Autograd. Therefore this function
+            only supports JAX, Torch and Tensorflow.
+
+        ..seealso:: `~.SpecialUnitary`
+
+        """
+        theta = self.data[0]
+        if len(qml.math.shape(theta)) > 1:
+            raise ValueError("Broadcasting is not supported.")
+
+        num_wires = self.hyperparameters["num_wires"]
+        if interface == "jax":
+            import jax
+
+            theta = qml.math.cast_like(theta, 1j)
+            jac = jax.jacobian(self.compute_matrix, argnums=0, holomorphic=True)(theta, num_wires)
+
+        elif interface in ["torch", "autograd"]:
+            # TODO check whether we can add support for Autograd using eigenvalue decomposition
+            if interface == "autograd":
+                raise NotImplementedError(
+                    "The matrix exponential expm is not differentiable in Autograd."
+                )
+
+            def real_matrix(theta):
+                return qml.math.real(self.compute_matrix(theta, num_wires))
+
+            def imag_matrix(theta):
+                return qml.math.imag(self.compute_matrix(theta, num_wires))
+
+            import torch
+
+            # These lines compute the Jacobian of compute_matrix every time -> to be optimized
+            real_jac = torch.autograd.functional.jacobian(real_matrix, theta)
+            imag_jac = torch.autograd.functional.jacobian(imag_matrix, theta)
+            jac = real_jac + 1j * imag_jac
+
+        elif interface in ("tensorflow", "tf"):
+            import tensorflow as tf
+
+            with tf.GradientTape(persistent=True) as tape:
+                mat = self.compute_matrix(theta, num_wires)
+                rmat, imat = tf.math.real(mat), tf.math.imag(mat)
```

Comment:
I thought more code could be reused, but at least it's an improvement.
```suggestion
        def real_matrix(theta):
            return qml.math.real(self.compute_matrix(theta, num_wires))

        def imag_matrix(theta):
            return qml.math.imag(self.compute_matrix(theta, num_wires))

        if interface == "jax":
            import jax

            theta = qml.math.cast_like(theta, 1j)
            jac = jax.jacobian(self.compute_matrix, argnums=0, holomorphic=True)(theta, num_wires)

        elif interface == "torch":
            import torch

            # These lines compute the Jacobian of compute_matrix every time -> to be optimized
            real_jac = torch.autograd.functional.jacobian(real_matrix, theta)
            imag_jac = torch.autograd.functional.jacobian(imag_matrix, theta)
            jac = real_jac + 1j * imag_jac

        elif interface in ("tensorflow", "tf"):
            import tensorflow as tf

            with tf.GradientTape(persistent=True) as tape:
                rmat, imat = real_matrix(theta), imag_matrix(theta)
```